File: ts-sdk/client/src/gpa/tokenBadge.ts
//
// Copyright (c) Cryptic Dot
//
// Modification based on Orca Whirlpools (https://github.com/orca-so/whirlpools),
// originally licensed under the Apache License, Version 2.0, prior to February 26, 2025.
//
// Modifications licensed under FusionAMM SDK Source-Available License v1.0
// See the LICENSE file in the project root for license information.
//

import type { GetProgramAccountsMemcmpFilter, Address, Account, GetProgramAccountsApi, Rpc } from "@solana/kit";
import { getBase58Decoder, getAddressEncoder } from "@solana/kit";
import type { TokenBadge } from "../generated/accounts/tokenBadge";
import { TOKEN_BADGE_DISCRIMINATOR, getTokenBadgeDecoder } from "../generated/accounts/tokenBadge";
import { fetchDecodedProgramAccounts } from "./utils";
import { FUSIONAMM_PROGRAM_ADDRESS } from "../generated/programs/fusionamm";

export type TokenBadgeFilter = GetProgramAccountsMemcmpFilter & {
  readonly __kind: unique symbol;
};

export function tokenBadgeFusionPoolsConfigFilter(address: Address): TokenBadgeFilter {
  return {
    memcmp: {
      offset: 8n,
      bytes: getBase58Decoder().decode(getAddressEncoder().encode(address)),
      encoding: "base58",
    },
  } as TokenBadgeFilter;
}

export function tokenBadgeTokenMintFilter(address: Address): TokenBadgeFilter {
  return {
    memcmp: {
      offset: 40n,
      bytes: getBase58Decoder().decode(getAddressEncoder().encode(address)),
      encoding: "base58",
    },
  } as TokenBadgeFilter;
}

export async function fetchAllTokenBadgeWithFilter(
  rpc: Rpc<GetProgramAccountsApi>,
  ...filters: TokenBadgeFilter[]
): Promise<Account<TokenBadge>[]> {
  const discriminator = getBase58Decoder().decode(TOKEN_BADGE_DISCRIMINATOR);
  const discriminatorFilter: GetProgramAccountsMemcmpFilter = {
    memcmp: {
      offset: 0n,
      bytes: discriminator,
      encoding: "base58",
    },
  };
  return fetchDecodedProgramAccounts(
    rpc,
    FUSIONAMM_PROGRAM_ADDRESS,
    [discriminatorFilter, ...filters],
    getTokenBadgeDecoder(),
  );
}

File: rust-sdk/client/src/pda/token_badge.rs
//
// Copyright (c) Cryptic Dot
//
// Modification based on Orca Whirlpools (https://github.com/orca-so/whirlpools),
// originally licensed under the Apache License, Version 2.0, prior to February 26, 2025.
//
// Modifications licensed under FusionAMM SDK Source-Available License v1.0
// See the LICENSE file in the project root for license information.
//

use crate::generated::programs::FUSIONAMM_ID;
use solana_program::program_error::ProgramError;
use solana_pubkey::Pubkey;

pub fn get_token_badge_address(token_mint: &Pubkey) -> Result<(Pubkey, u8), ProgramError> {
    let seeds = &[b"token_badge", token_mint.as_ref()];
    Pubkey::try_find_program_address(seeds, &FUSIONAMM_ID).ok_or(ProgramError::InvalidSeeds)
}
